### PR TITLE
String distance, first skeleton

### PR DESCRIPTION
**Pull request by quantumFeline**

### CODE REVIEW

File: gTea/app/src/main/java/com/nechaieva/gtea/OrderProcessor.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
+package com.nechaieva.gtea;
+
+import java.lang.reflect.Array;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+public class OrderProcessor {
```
Comment (AlexanderFroemmgen) [inline gTea/app/src/main/java/com/nechaieva/gtea/OrderProcessor.java]: Consider some comments/documentation what this class is supposed to do

File: gTea/app/src/main/java/com/nechaieva/gtea/OrderProcessor.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+    List<String> menu;
```
Comment (AlexanderFroemmgen) [inline gTea/app/src/main/java/com/nechaieva/gtea/OrderProcessor.java]: Consider adding a visibility qualifier

File: gTea/app/src/main/java/com/nechaieva/gtea/OrderProcessor.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    final int SENSITIVITY = 3;
```
Comment (AlexanderFroemmgen) [inline gTea/app/src/main/java/com/nechaieva/gtea/OrderProcessor.java]: Consider adding a one line comment per attribute, e.g., what is this sensitivity doing? is it the logging sensitivity?

File: gTea/app/src/main/java/com/nechaieva/gtea/OrderProcessor.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    public final String NO_CANDIDATE = "";
+
+    OrderProcessor(String[] menu) {
+        this.menu = Arrays.asList(menu);
+    }
+
+    OrderProcessor(ArrayList<String> menu) {
```
Comment (AlexanderFroemmgen) [inline gTea/app/src/main/java/com/nechaieva/gtea/OrderProcessor.java]: Consider adding a short comment for each method.

File: gTea/app/src/main/java/com/nechaieva/gtea/OrderProcessor.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        this.menu = menu;
```
Comment (AlexanderFroemmgen) [inline gTea/app/src/main/java/com/nechaieva/gtea/OrderProcessor.java]: Is it intentionally that you don't copy the underlying list? This seems to be an interesting design decision and I would recommend to add a comment about your reasoning for this decision.

File: gTea/app/src/main/java/com/nechaieva/gtea/OrderProcessor.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    public String findInMenu(String query) {
+        Match match = findClosest(query);
+        return match.getDistance() < SENSITIVITY ? match.getMatchString() : NO_CANDIDATE;
```
Comment (AlexanderFroemmgen) [inline gTea/app/src/main/java/com/nechaieva/gtea/OrderProcessor.java]: What about return an optional instead? This would make the "no result" case more expressive...

**Review by AlexanderFroemmgen — CHANGES_REQUESTED** [review]
Thanks a lot.

What about adding some "still failing" tests early?

File: gTea/app/src/main/java/com/nechaieva/gtea/OrderProcessor.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    private Match findClosest(String query) {
+        // Probably a better approach than a straightforward iterate-through exists
```
Comment (wefelix) [inline gTea/app/src/main/java/com/nechaieva/gtea/OrderProcessor.java]: As you have to compute the distance for all elements in the list at least once, you have to iterate through it at least once in the worst case. You could consider adding optimisation heuristics e.g. breaking early if you have a zero distance match.

**Review by wefelix — APPROVED** [review]
Looks good! Having a separate class for the order processor is elegant, please add a test class.

File: gTea/app/src/main/java/com/nechaieva/gtea/OrderProcessor.java (hunk 7 above)
Comment (AlexanderFroemmgen) [reply gTea/app/src/main/java/com/nechaieva/gtea/OrderProcessor.java]: Nice idea. Breaking in case there was already a better match would be possible as well ;-)

I like this Knuth citation about optimization

File: gTea/app/src/main/java/com/nechaieva/gtea/OrderProcessor.java [hunk 8]
```diff
@@ -1,15 +1,19 @@
 package com.nechaieva.gtea;
 
-import java.lang.reflect.Array;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.List;
 
+/**
+ * Given a query that represent the user's order,
+ * tries to process the order so that it can be passed back to the handler.
```
Comment (AlexanderFroemmgen) [inline gTea/app/src/main/java/com/nechaieva/gtea/OrderProcessor.java]: I would propose to describe what the result is but not make concrete assumptions about the concrete user of the handler. Imagine you are just writing this functionality to be reused...

File: gTea/app/src/main/java/com/nechaieva/gtea/OrderProcessor.java [hunk 9]
```diff
@@ -1,15 +1,19 @@
...
+ * For this, it is necessary to find to what kind of menu item / items
+ * it corresponds.
+ */
 public class OrderProcessor {
 
-    List<String> menu;
-    final int SENSITIVITY = 3;
-    public final String NO_CANDIDATE = "";
+    private List<String> menu;
+    public final int DEFAULT_SENSITIVITY = 3; // The threshold for string distances.
```
Comment (AlexanderFroemmgen) [inline gTea/app/src/main/java/com/nechaieva/gtea/OrderProcessor.java]: Can you be more precise, e.g., is it considered okay to have 3?

Comment (AlexanderFroemmgen) [reply gTea/app/src/main/java/com/nechaieva/gtea/OrderProcessor.java]: See offline discussion for more background

File: gTea/app/src/main/java/com/nechaieva/gtea/OrderProcessor.java [hunk 10]
```diff
@@ -19,11 +23,39 @@
...
+     * @see OrderProcessor#findInMenu(String, String, int)
+     * Uses default sensitivity.
+     */
+    public String findInMenu(String query,
+                             String defaultNoCandidate) {
+        return findInMenu(query, defaultNoCandidate, DEFAULT_SENSITIVITY);
+    }
+
+    /**
+     * Given a text query passed from the user, return the string that represent
```
Comment (AlexanderFroemmgen) [inline gTea/app/src/main/java/com/nechaieva/gtea/OrderProcessor.java]: Same as above. Consider focusing on the "what to expect from this method" rather than "the architecture" of your current setup.

File: gTea/app/src/main/java/com/nechaieva/gtea/OrderProcessor.java [hunk 11]
```diff
@@ -39,12 +71,15 @@ private Match findClosest(String query) {
         return new Match(bestMatch, minStringDistance);
     }
 
+    /**
+     * Calculates the distance between two string, using Levenshtein algorithm.
```
Comment (AlexanderFroemmgen) [inline gTea/app/src/main/java/com/nechaieva/gtea/OrderProcessor.java]: I'm not sure if I added this comment here or for Margaret, but it seems like the method name and the comment are not synced... If the method is supposed to only use the Levenshtein algorithm, what about naming the method accordingly?

File: gTea/app/src/main/java/com/nechaieva/gtea/BaseOrderProcessor.java [hunk 12]
```diff
@@ -0,0 +1,29 @@
+package com.nechaieva.gtea;
+
+import java.util.Arrays;
+import java.util.Optional;
+
+/**
+ * A base implementation of OrderProcessor.
+ * Fast but only works with exact queries and does not understand fuzzy ones.
+ */
+public class BaseOrderProcessor implements OrderProcessor {
```
Comment (AlexanderFroemmgen) [inline gTea/app/src/main/java/com/nechaieva/gtea/BaseOrderProcessor.java]: Did you consider to name it "ExactMatchOrderProcessor"?

File: gTea/app/src/main/java/com/nechaieva/gtea/LevenshteinOrderProcessor.java [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * And implementation of OrderProcessor that uses Levenshtein distances.
+ */
+public class LevenshteinOrderProcessor implements OrderProcessor {
```
Comment (AlexanderFroemmgen) [inline gTea/app/src/main/java/com/nechaieva/gtea/LevenshteinOrderProcessor.java]: FYI: Some coding styles encode the "implementation vs interface" in the name...

File: gTea/app/src/main/java/com/nechaieva/gtea/OrderProcessor.java [hunk 14]
```diff
@@ -1,99 +1,29 @@
...
-
-        public String getMatchString() {
-            return matchString;
-        }
-
-        public int getDistance() {
-            return distance;
-        }
-    }
+    Optional<String> findInMenu(String query);
```
Comment (AlexanderFroemmgen) [inline gTea/app/src/main/java/com/nechaieva/gtea/OrderProcessor.java]: What happens if findInMenu is called before loadMenu? This might be a nice topic for a more detailed interactive discussion...

File: gTea/app/src/main/java/com/nechaieva/gtea/LevenshteinDistCalculator.java [hunk 15]
```diff
@@ -0,0 +1,12 @@
+package com.nechaieva.gtea;
+
+public class LevenshteinDistCalculator {
```
Comment (AlexanderFroemmgen) [inline gTea/app/src/main/java/com/nechaieva/gtea/LevenshteinDistCalculator.java]: Consider a private constructor if you don't expect any instances of this class

File: gTea/app/src/main/java/com/nechaieva/gtea/LevenshteinOrderProcessor.java [hunk 16]
```diff
@@ -32,6 +33,9 @@ public void loadMenu(String[] menu) {
      */
     @Override
     public Optional<String> findInMenu(String query) {
+        if(query.length() > maxAllowedStringLength) {
```
Comment (AlexanderFroemmgen) [inline gTea/app/src/main/java/com/nechaieva/gtea/LevenshteinOrderProcessor.java]: Consider documenting this behavior in the method description.

File: gTea/app/src/main/java/com/nechaieva/gtea/MenuFragment.java [hunk 17]
```diff
@@ -14,6 +14,13 @@
 import androidx.recyclerview.widget.LinearLayoutManager;
 import androidx.recyclerview.widget.RecyclerView;
 
+import com.nechaieva.gtea.DeepLink;
```
Comment (AlexanderFroemmgen) [inline gTea/app/src/main/java/com/nechaieva/gtea/MenuFragment.java:17]: Are these changes due to the Java version change?

Consider being consistent with the commit description/name

File: gTea/app/src/main/java/com/nechaieva/gtea/utils/BaseOrderProcessor.java [hunk 18]
```diff
@@ -23,7 +23,14 @@ public void loadMenu(String[] menu) {
 
     @Override
     public Optional<String> findInMenu(String query) {
+        query = normalize(query);
```
Comment (AlexanderFroemmgen) [inline gTea/app/src/main/java/com/nechaieva/gtea/utils/BaseOrderProcessor.java]: As mentioned in the other commit, consider being consistent with the name/description of the commit

File: gTea/app/src/test/java/com/nechaieva/gtea/BaseOrderProcessorTest.java [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+    public static void setUp() throws Exception {
+        String[] menu = {"Black tea", "Green tea", "Coffee", "Donuts"};
+        processor = new BaseOrderProcessor(menu);
+    }
+
+    @Test
+    public void findExistent() {
+        String userQuery = "Black tea";
+        Optional<String> res = processor.findInMenu(userQuery);
+        assertEquals(res.orElse("null"), "black tea");
```
Comment (AlexanderFroemmgen) [inline gTea/app/src/test/java/com/nechaieva/gtea/BaseOrderProcessorTest.java]: Is there a more expressive way to handle optional's in tests?

File: gTea/app/src/test/java/com/nechaieva/gtea/BaseOrderProcessorTest.java [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
```
Comment (AlexanderFroemmgen) [inline gTea/app/src/test/java/com/nechaieva/gtea/BaseOrderProcessorTest.java]: Are there additional relevant corner cases?

File: gTea/app/src/main/java/com/nechaieva/gtea/utils/Normalizer.java [hunk 21]
```diff
@@ -1,4 +1,20 @@
-package com.nechaieva.gtea;
+package com.nechaieva.gtea.utils;
 
+/**
+ * A class used to normalize strings for the order.
```
Comment (AlexanderFroemmgen) [inline gTea/app/src/main/java/com/nechaieva/gtea/utils/Normalizer.java:4]: What about being a bit more precise on the expected behavior, e.g., what it does?

**Review by AlexanderFroemmgen — COMMENTED** [review]
Thanks, I did an initial iteration.

File: gTea/app/src/test/java/com/nechaieva/gtea/BaseOrderProcessorTest.java (hunk 20 above)
Comment (quantumFeline) [reply gTea/app/src/test/java/com/nechaieva/gtea/BaseOrderProcessorTest.java]: What kind of cases do you have in mind?

File: gTea/app/src/main/java/com/nechaieva/gtea/utils/LevenshteinDistCalculator.java [hunk 22]
```diff
@@ -1,12 +1,45 @@
 package com.nechaieva.gtea.utils;
 
 public class LevenshteinDistCalculator {
+    static int[][] dp;
```
Comment (AlexanderFroemmgen) [inline gTea/app/src/main/java/com/nechaieva/gtea/utils/LevenshteinDistCalculator.java:8]: What visibility is needed here?

File: gTea/app/src/main/java/com/nechaieva/gtea/utils/LevenshteinDistCalculator.java [hunk 23]
```diff
@@ -1,12 +1,45 @@
 package com.nechaieva.gtea.utils;
 
 public class LevenshteinDistCalculator {
```
Comment (AlexanderFroemmgen) [inline gTea/app/src/main/java/com/nechaieva/gtea/utils/LevenshteinDistCalculator.java:3]: Could you please add some comments to this class? While the code might be correct, it's hard to understand without a lot of time investment. Some comments and pointers to "inspirations", e.g., the name of the used algorithm in case it's a well known one, might simplify life of reviewers and future readers of this code significantly.

File: gTea/app/src/main/java/com/nechaieva/gtea/utils/LevenshteinDistCalculator.java [hunk 24]
```diff
@@ -1,12 +1,45 @@
...
+        }
+        for (int y = 1; y < al; y++) {
+            for (int x = 1; x < bl; x++) {
+                int charEquals = a.charAt(y) != b.charAt(x) ? 1 : 0;
+                dp[y][x] = min(dp[y][x-1] + 1,
+                               dp[y-1][x] + 1,
+                               dp[y-1][x-1] + charEquals);
+            }
+        }
+        printArray(al, bl);
```
Comment (AlexanderFroemmgen) [inline gTea/app/src/main/java/com/nechaieva/gtea/utils/LevenshteinDistCalculator.java]: It feels like the print part is a kind of "test logging"... did you consider to use a logger for this?

File: gTea/app/src/main/java/com/nechaieva/gtea/utils/LevenshteinDistCalculator.java [hunk 25]
```diff
@@ -1,12 +1,45 @@
...
+        return dp[al-1][bl-1];
+    }
+
+    private static int min(int a, int b, int c) {
+        return Math.min(Math.min(a, b), c);
+    }
+
+    private static void printArray(int ySize, int xSize) {
```
Comment (AlexanderFroemmgen) [inline gTea/app/src/main/java/com/nechaieva/gtea/utils/LevenshteinDistCalculator.java:54]: What array is this printing? What's are the parameters used for?

File: gTea/app/src/test/java/com/nechaieva/gtea/utils/LevenshteinDistCalculatorTest.java [hunk 26]
```diff
@@ -0,0 +1,32 @@
...
+    public void testLetterChanged() {
+        test( 1,"hell", "hall");
+    }
+
+    @Test
+    public void testClassicalExample() {
+        test( 6,"polynomial", "exponential");
+    }
+
+    private void test(int dist, String a, String b) {
```
Comment (AlexanderFroemmgen) [inline gTea/app/src/test/java/com/nechaieva/gtea/utils/LevenshteinDistCalculatorTest.java:34]: I'm not sure about the purpose/usefulness of this function. It feels like removing the function and using assert in the Testcases makes the code easier to read. WDYT?

File: gTea/app/src/main/java/com/nechaieva/gtea/utils/LevenshteinDistCalculator.java [hunk 27]
```diff
@@ -5,6 +5,7 @@
 
     /**
      * Calculates given distance between two given strings.
+     * Currently, uses the Levenshtein-Damerau version of the algorithm.
```
Comment (AlexanderFroemmgen) [inline gTea/app/src/main/java/com/nechaieva/gtea/utils/LevenshteinDistCalculator.java:12]: "Currently"... what the purpose and implications of this statement? Imagine you are going to reuse an existing method which mentions in the comment what they are currently doing.... Do you feel like it's safe for you to rely on this method?

Comment (AlexanderFroemmgen) [inline gTea/app/src/main/java/com/nechaieva/gtea/utils/LevenshteinDistCalculator.java:12]: Did you consider alternatives such as an additional class to implement the new algorithm or a configuration parameter to choose between the different algorithms instead of fully replacing the existing code?

Comment (AlexanderFroemmgen) [inline gTea/app/src/main/java/com/nechaieva/gtea/utils/LevenshteinDistCalculator.java:12]: Did you consider to add an additional sentence or a few words to explain what the Levenshtein-Damerau algorithm is?

File: gTea/app/src/test/java/com/nechaieva/gtea/utils/LevenshteinDistCalculatorTest.java [hunk 28]
```diff
@@ -21,6 +21,11 @@ public void testLetterChanged() {
         test( 1,"hell", "hall");
     }
 
+    @Test
```
Comment (AlexanderFroemmgen) [inline gTea/app/src/test/java/com/nechaieva/gtea/utils/LevenshteinDistCalculatorTest.java:24]: Writing tests and having proper test coverage is a hard task.

This concrete example is a nice illustration that you probably did not have proper test coverage before as no of the existing tests breaks with the new algorithm.

File: gTea/app/src/test/java/com/nechaieva/gtea/utils/LevenshteinDistCalculatorTest.java [hunk 29]
```diff
@@ -0,0 +1,32 @@
+package com.nechaieva.gtea.utils;
+
+import org.junit.Test;
+
+import static org.junit.Assert.*;
+
+public class LevenshteinDistCalculatorTest {
```
Comment (AlexanderFroemmgen) [inline gTea/app/src/test/java/com/nechaieva/gtea/utils/LevenshteinDistCalculatorTest.java:7]: Tests should try to capture all relevant cases. This includes "corner cases", e.g., does it make a difference if the letter which was changed is a the beginning, in the middle, or the end of the string? Does your algorithm work for output values bigger than X as well? what happens with non-ascii input?

File: gTea/app/src/test/java/com/nechaieva/gtea/utils/LevenshteinDistCalculatorTest.java [hunk 30]
```diff
@@ -0,0 +1,32 @@
...
+
+    @Test
+    public void testLetterMissing() {
+        test( 1,"hello", "hell");
```
Comment (AlexanderFroemmgen) [inline gTea/app/src/test/java/com/nechaieva/gtea/utils/LevenshteinDistCalculatorTest.java:11]: There are some formatting inconsistencies. Please fix.

File: gTea/app/src/main/java/com/nechaieva/gtea/fragments/MenuFragment.java [hunk 31]
```diff
@@ -112,12 +115,15 @@ void checkVoiceOrder() {
             Log.e(ORDER_TAG, "No MainActivity found");
             return;
         }
+
         Bundle order = mainActivity.getStartingInfo();
+        Optional<String> menuItem = checkInMenu(order);
```
Comment (AlexanderFroemmgen) [inline gTea/app/src/main/java/com/nechaieva/gtea/fragments/MenuFragment.java:120]: It feels like this code could be refactored to simplify it.

Is the variable "navigateTo" required?

Proposal:

if (!menuItem.IsPresent()) {
   // some logging
  return;
}
order.putString(DeepLink.ORDER_BUNDLE_TAG, menuItem.get());
navigateToOrder(order); // put the success logging into this method

File: gTea/app/src/main/java/com/nechaieva/gtea/fragments/MenuFragment.java [hunk 32]
```diff
@@ -142,4 +148,15 @@ void navigateToOrder(Bundle bundle) {
...
     }
+
+    Optional<String> checkInMenu(Bundle order) {
+        if (order == null || order.isEmpty() || order.getString(DeepLink.ORDER_BUNDLE_TAG) == null) {
+            return Optional.empty();
+        }
+        return findMenuItem(order.getString(DeepLink.ORDER_BUNDLE_TAG));
+    }
+
+    Optional<String> findMenuItem(String query) {
```
Comment (AlexanderFroemmgen) [inline gTea/app/src/main/java/com/nechaieva/gtea/fragments/MenuFragment.java:159]: It this method needed?

File: gTea/app/src/main/java/com/nechaieva/gtea/utils/LevenshteinDistCalculator.java (hunk 23 above)
Comment (AlexanderFroemmgen) [inline gTea/app/src/main/java/com/nechaieva/gtea/utils/LevenshteinDistCalculator.java:3]: If you only have static methods here, did you consider to have a private constructor to avoid instanciations?

File: gTea/app/src/main/java/com/nechaieva/gtea/utils/LevenshteinDistCalculator.java [hunk 33]
```diff
@@ -1,6 +1,10 @@
 package com.nechaieva.gtea.utils;
 
 public class LevenshteinDistCalculator {
+
+    private LevenshteinDistCalculator() {
```
Comment (AlexanderFroemmgen) [inline gTea/app/src/main/java/com/nechaieva/gtea/utils/LevenshteinDistCalculator.java:5]: This is a great example (again!) that our current review process with the single simple cl's does not work. I flagged it while going through your cl even though you addressed it already in a future cl....

File: gTea/app/src/test/java/com/nechaieva/gtea/utils/LevenshteinOrderProcessorTest.java [hunk 34]
```diff
@@ -0,0 +1,52 @@
+package com.nechaieva.gtea.utils;
```
Comment (AlexanderFroemmgen) [inline gTea/app/src/test/java/com/nechaieva/gtea/utils/LevenshteinOrderProcessorTest.java:1]: I would propose to have code and tests in a single commit in the future

File: gTea/app/src/main/java/com/nechaieva/gtea/utils/LevenshteinOrderProcessor.java [hunk 35]
```diff
@@ -11,7 +11,13 @@
 public class LevenshteinOrderProcessor implements OrderProcessor {
 
     private List<String> menu;
-    public final int DEFAULT_SENSITIVITY = 3; // The threshold for string distances.
+
+    /**
+     * The threshold for string distances.
+     * Represents the number of typos the user can make before the query will be rejected.
```
Comment (AlexanderFroemmgen) [inline gTea/app/src/main/java/com/nechaieva/gtea/utils/LevenshteinOrderProcessor.java:17]: "number of typos" seems rather imprecise given that there is. a more precise definition available

File: gTea/app/src/main/java/com/nechaieva/gtea/utils/LevenshteinOrderProcessor.java [hunk 36]
```diff
@@ -58,7 +61,17 @@ public void loadMenu(ArrayList<String> menu) {
      * A menu item.
      */
     public Optional<String> findInMenu(String query, int sensitivity) {
+
+        if (menu == null) {
+            throw new NullPointerException("Menu is not initialized");
```
Comment (AlexanderFroemmgen) [inline gTea/app/src/main/java/com/nechaieva/gtea/utils/LevenshteinOrderProcessor.java:66]: Did you consider alternative exception types?

Did you consider alternative designs which avoid this failure case at all?

File: gTea/app/src/main/java/com/nechaieva/gtea/utils/LevenshteinDistCalculator.java [hunk 37]
```diff
@@ -37,14 +37,20 @@ static int stringDistance(String a, String b) {
                 }
             }
         }
-        printArray(al, bl);
```
Comment (AlexanderFroemmgen) [inline gTea/app/src/main/java/com/nechaieva/gtea/utils/LevenshteinDistCalculator.java]: Who is calling this method now?

**Review by AlexanderFroemmgen — CHANGES_REQUESTED** [review]
Thanks a lot, I did one iteration on the new cl's

File: gTea/app/src/main/java/com/nechaieva/gtea/utils/LevenshteinDistCalculator.java (hunk 23 above)
Comment (quantumFeline) [reply gTea/app/src/main/java/com/nechaieva/gtea/utils/LevenshteinDistCalculator.java:3]: Currently, stringDistance() docstring mentions the algorithm name. Is there anything else you think I should add?

File: gTea/app/src/main/java/com/nechaieva/gtea/utils/LevenshteinDistCalculator.java (hunk 27 above)
Comment (quantumFeline) [reply gTea/app/src/main/java/com/nechaieva/gtea/utils/LevenshteinDistCalculator.java:12]: That was my first thought as well. However, I feel like that the new version is strictly better for the scenarios in which it is used, and I didn't want to over-engineer.
